File: python/demo13a_mutex.py
'''
MUTEXES
In Python, Lock objects can be used as mutexes
'''

import time
import threading



mutex = threading.Lock()
counter = 0



def do_task():
    global counter

    mutex.acquire()

    for _ in range(1000):
        temp = counter + 1
        time.sleep(0.0001)
        counter = temp

    mutex.release()



NUM_THREADS = 32
lstth = [threading.Thread(target=do_task) for _ in range(NUM_THREADS)]

for th in lstth:
    th.start()

for th in lstth:
    th.join()

print('counter =', counter)
# We are sure that counter = 32000

File: cpp/cpp-std/demo13a-mutex.cpp
/*
MUTEXES
*/


#include <iostream>
#include <chrono>
#include <thread>
#include <mutex>
using namespace std;



std::mutex mut;
int counter = 0;



void doTask() {
    std::this_thread::sleep_for(std::chrono::seconds(1));

    mut.lock();

    for (int i = 0; i < 1000; ++i)
        ++counter;

    mut.unlock();
}



int main() {
    constexpr int NUM_THREADS = 16;
    std::thread lstTh[NUM_THREADS];

    for (auto&& th : lstTh) {
        th = std::thread(doTask);
    }

    for (auto&& th : lstTh) {
        th.join();
    }

    cout << "counter = " << counter << endl;
    // We are sure that counter = 16000

    return 0;
}
